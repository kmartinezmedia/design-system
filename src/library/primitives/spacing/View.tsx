import React, { FunctionComponent } from 'react';
import {
  View as RNView,
  ViewStyle,
  ViewProps as RNViewProps,
  FlexStyle,
  GestureResponderEvent,
  Animated,
  StyleSheet,
} from 'react-native';
import { useSpacing, usePin, useElevation } from '@designSystem/hooks';
import {
  Spacing,
  PinningDirection,
  ElevationLevels,
} from '@designSystem/types';
import LinearGradient from 'react-native-linear-gradient';
import { useTheme } from '@designSystem/theme';

export interface ViewProps
  extends Omit<RNViewProps, 'style'>,
    Pick<
      FlexStyle,
      | 'height'
      | 'width'
      | 'minWidth'
      | 'flexGrow'
      | 'flexShrink'
      | 'justifyContent'
      | 'alignItems'
      | 'alignContent'
      | 'alignSelf'
      | 'flexDirection'
      | 'flexWrap'
      | 'flexBasis'
    > {
  spacing?: Spacing;
  dangerouslySetStyle?: ViewStyle;
  elevation?: ElevationLevels;
  pin?: PinningDirection;
  onPress?: (event: GestureResponderEvent) => boolean;
  animated?: boolean;
  overflow?: 'visible' | 'hidden' | 'scroll' | 'gradient';
}

export const View: FunctionComponent<ViewProps> = ({
  overflow = 'visible',
  flexDirection = 'column',
  flexWrap = 'nowrap',
  justifyContent = 'flex-start',
  alignItems = 'stretch',
  alignSelf = 'auto',
  spacing = 0,
  dangerouslySetStyle = {},
  flexGrow,
  flexShrink,
  height,
  width,
  animated,
  elevation,
  pin,
  onPress,
  children,
  ...props
}) => {
  const { colors } = useTheme();
  const spacingStyles = useSpacing(spacing);
  const pinStyles = usePin(pin);
  const elevationStyles = useElevation(elevation);

  const Component = animated ? Animated.View : RNView;

  return (
    <Component
      onStartShouldSetResponder={onPress}
      style={
        {
          ...{
            flexDirection,
            flexWrap,
            justifyContent,
            alignItems,
            alignSelf,
            flexGrow,
            flexShrink,
            height,
            width,
          },
          ...spacingStyles,
          ...pinStyles,
          ...elevationStyles,
          ...dangerouslySetStyle,
        } as any
      }
      {...props}>
      {children}
      {overflow === 'gradient' ? (
        <LinearGradient
          style={styles.Gradient}
          colors={[
            '#ffffff00' /* oneoff color bc android can't gradient from 'transparent' */,
            colors.background.default,
          ]}
          start={{ x: 0, y: 0 }}
          end={{ x: 1, y: 0 }}
        />
      ) : null}
    </Component>
  );
};

const styles = StyleSheet.create({
  Gradient: {
    width: 30,
    top: 0,
    right: 0,
    bottom: 0,
    position: 'absolute' as const,
  },
});
