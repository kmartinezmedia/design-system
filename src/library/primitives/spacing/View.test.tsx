import React from 'react';
import 'react-native';
import { create, act, ReactTestRenderer } from 'react-test-renderer';
import { View } from './View';

describe('View', () => {
  it('default', () => {
    let component: ReactTestRenderer;
    act(() => {
      component = create(<View />);
    });
    expect(component!.toJSON()).toMatchSnapshot();
  });
  it('spacing', () => {
    let component: ReactTestRenderer;
    act(() => {
      component = create(
        <>
          {/* string */}
          <View spacing={0} />
          <View spacing={6} />

          {/* shorthand */}
          <View spacing={[0]} />
          <View spacing={[6]} />
          <View spacing={[0, 4]} />
          <View spacing={[0, -4, 4]} />
          <View spacing={[0, -4, 4, 4]} />

          {/* object */}
          <View spacing={{ top: 4 }} />
          <View spacing={{ horizontal: 4 }} />
          <View spacing={{ horizontal: 4, bottom: -3 }} />
          <View spacing={{ horizontal: 4, vertical: 3 }} />
          <View spacing={{ top: 4, bottom: 3, left: 0, right: -6 }} />
        </>,
      );
    });
    expect(component!.toJSON()).toMatchSnapshot();
  });
  it('dangerouslySetStyle', () => {
    let component: ReactTestRenderer;
    const styles = { backgroundColor: 'red' };
    act(() => {
      component = create(<View dangerouslySetStyle={styles} />);
    });
    expect(component!.toJSON()).toMatchSnapshot();
  });
  it('elevation', () => {
    let component: ReactTestRenderer;
    act(() => {
      component = create(
        <>
          <View elevation={1} />
          <View elevation={2} />
        </>,
      );
    });
    expect(component!.toJSON()).toMatchSnapshot();
  });
  it('pin', () => {
    let component: ReactTestRenderer;
    act(() => {
      component = create(
        <>
          <View pin="left" />
          <View pin="top" />
          <View pin="right" />
          <View pin="bottom" />
        </>,
      );
    });
    expect(component!.toJSON()).toMatchSnapshot();
  });
});
