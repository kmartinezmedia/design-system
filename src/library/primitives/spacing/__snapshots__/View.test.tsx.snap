// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View dangerouslySetStyle 1`] = `
<View
  style={
    Object {
      "alignItems": "stretch",
      "alignSelf": "auto",
      "backgroundColor": "red",
      "flexDirection": "column",
      "flexGrow": undefined,
      "flexShrink": undefined,
      "flexWrap": "nowrap",
      "height": undefined,
      "justifyContent": "flex-start",
      "width": undefined,
    }
  }
/>
`;

exports[`View default 1`] = `
<View
  style={
    Object {
      "alignItems": "stretch",
      "alignSelf": "auto",
      "flexDirection": "column",
      "flexGrow": undefined,
      "flexShrink": undefined,
      "flexWrap": "nowrap",
      "height": undefined,
      "justifyContent": "flex-start",
      "width": undefined,
    }
  }
/>
`;

exports[`View elevation 1`] = `
Array [
  <View
    style={
      Object {
        "alignItems": "stretch",
        "alignSelf": "auto",
        "borderColor": "#F1F3F5",
        "borderWidth": 1,
        "flexDirection": "column",
        "flexGrow": undefined,
        "flexShrink": undefined,
        "flexWrap": "nowrap",
        "height": undefined,
        "justifyContent": "flex-start",
        "shadowColor": "#071422",
        "shadowOffset": Object {
          "height": 8,
          "width": 0,
        },
        "shadowOpacity": 0.02,
        "shadowRadius": 12,
        "width": undefined,
      }
    }
  />,
  <View
    style={
      Object {
        "alignItems": "stretch",
        "alignSelf": "auto",
        "borderColor": "#F1F3F5",
        "borderWidth": 1,
        "flexDirection": "column",
        "flexGrow": undefined,
        "flexShrink": undefined,
        "flexWrap": "nowrap",
        "height": undefined,
        "justifyContent": "flex-start",
        "shadowColor": "#071422",
        "shadowOffset": Object {
          "height": 8,
          "width": 0,
        },
        "shadowOpacity": 0.24,
        "shadowRadius": 40,
        "width": undefined,
      }
    }
  />,
]
`;

exports[`View pin 1`] = `
Array [
  <View
    style={
      Object {
        "alignItems": "stretch",
        "alignSelf": "auto",
        "bottom": 0,
        "flexDirection": "column",
        "flexGrow": undefined,
        "flexShrink": undefined,
        "flexWrap": "nowrap",
        "height": undefined,
        "justifyContent": "flex-start",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": undefined,
      }
    }
  />,
  <View
    style={
      Object {
        "alignItems": "stretch",
        "alignSelf": "auto",
        "flexDirection": "column",
        "flexGrow": undefined,
        "flexShrink": undefined,
        "flexWrap": "nowrap",
        "height": undefined,
        "justifyContent": "flex-start",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "width": undefined,
      }
    }
  />,
  <View
    style={
      Object {
        "alignItems": "stretch",
        "alignSelf": "auto",
        "bottom": 0,
        "flexDirection": "column",
        "flexGrow": undefined,
        "flexShrink": undefined,
        "flexWrap": "nowrap",
        "height": undefined,
        "justifyContent": "flex-start",
        "position": "absolute",
        "right": 0,
        "top": 0,
        "width": undefined,
      }
    }
  />,
  <View
    style={
      Object {
        "alignItems": "stretch",
        "alignSelf": "auto",
        "bottom": 0,
        "flexDirection": "column",
        "flexGrow": undefined,
        "flexShrink": undefined,
        "flexWrap": "nowrap",
        "height": undefined,
        "justifyContent": "flex-start",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "width": undefined,
      }
    }
  />,
]
`;

exports[`View spacing 1`] = `
Array [
  <View
    style={
      Object {
        "alignItems": "stretch",
        "alignSelf": "auto",
        "flexDirection": "column",
        "flexGrow": undefined,
        "flexShrink": undefined,
        "flexWrap": "nowrap",
        "height": undefined,
        "justifyContent": "flex-start",
        "width": undefined,
      }
    }
  />,
  <View
    style={
      Object {
        "alignItems": "stretch",
        "alignSelf": "auto",
        "flexDirection": "column",
        "flexGrow": undefined,
        "flexShrink": undefined,
        "flexWrap": "nowrap",
        "height": undefined,
        "justifyContent": "flex-start",
        "padding": 24,
        "width": undefined,
      }
    }
  />,
  <View
    style={
      Object {
        "alignItems": "stretch",
        "alignSelf": "auto",
        "flexDirection": "column",
        "flexGrow": undefined,
        "flexShrink": undefined,
        "flexWrap": "nowrap",
        "height": undefined,
        "justifyContent": "flex-start",
        "padding": 0,
        "width": undefined,
      }
    }
  />,
  <View
    style={
      Object {
        "alignItems": "stretch",
        "alignSelf": "auto",
        "flexDirection": "column",
        "flexGrow": undefined,
        "flexShrink": undefined,
        "flexWrap": "nowrap",
        "height": undefined,
        "justifyContent": "flex-start",
        "padding": 24,
        "width": undefined,
      }
    }
  />,
  <View
    style={
      Object {
        "alignItems": "stretch",
        "alignSelf": "auto",
        "flexDirection": "column",
        "flexGrow": undefined,
        "flexShrink": undefined,
        "flexWrap": "nowrap",
        "height": undefined,
        "justifyContent": "flex-start",
        "paddingHorizontal": 16,
        "paddingVertical": 0,
        "width": undefined,
      }
    }
  />,
  <View
    style={
      Object {
        "alignItems": "stretch",
        "alignSelf": "auto",
        "flexDirection": "column",
        "flexGrow": undefined,
        "flexShrink": undefined,
        "flexWrap": "nowrap",
        "height": undefined,
        "justifyContent": "flex-start",
        "marginHorizontal": -16,
        "paddingBottom": 16,
        "paddingTop": 0,
        "width": undefined,
      }
    }
  />,
  <View
    style={
      Object {
        "alignItems": "stretch",
        "alignSelf": "auto",
        "flexDirection": "column",
        "flexGrow": undefined,
        "flexShrink": undefined,
        "flexWrap": "nowrap",
        "height": undefined,
        "justifyContent": "flex-start",
        "marginRight": -16,
        "paddingBottom": 16,
        "paddingLeft": 16,
        "paddingTop": 0,
        "width": undefined,
      }
    }
  />,
  <View
    style={
      Object {
        "alignItems": "stretch",
        "alignSelf": "auto",
        "flexDirection": "column",
        "flexGrow": undefined,
        "flexShrink": undefined,
        "flexWrap": "nowrap",
        "height": undefined,
        "justifyContent": "flex-start",
        "paddingTop": 16,
        "width": undefined,
      }
    }
  />,
  <View
    style={
      Object {
        "alignItems": "stretch",
        "alignSelf": "auto",
        "flexDirection": "column",
        "flexGrow": undefined,
        "flexShrink": undefined,
        "flexWrap": "nowrap",
        "height": undefined,
        "justifyContent": "flex-start",
        "paddingHorizontal": 16,
        "width": undefined,
      }
    }
  />,
  <View
    style={
      Object {
        "alignItems": "stretch",
        "alignSelf": "auto",
        "flexDirection": "column",
        "flexGrow": undefined,
        "flexShrink": undefined,
        "flexWrap": "nowrap",
        "height": undefined,
        "justifyContent": "flex-start",
        "marginBottom": -12,
        "paddingHorizontal": 16,
        "width": undefined,
      }
    }
  />,
  <View
    style={
      Object {
        "alignItems": "stretch",
        "alignSelf": "auto",
        "flexDirection": "column",
        "flexGrow": undefined,
        "flexShrink": undefined,
        "flexWrap": "nowrap",
        "height": undefined,
        "justifyContent": "flex-start",
        "paddingHorizontal": 16,
        "paddingVertical": 12,
        "width": undefined,
      }
    }
  />,
  <View
    style={
      Object {
        "alignItems": "stretch",
        "alignSelf": "auto",
        "flexDirection": "column",
        "flexGrow": undefined,
        "flexShrink": undefined,
        "flexWrap": "nowrap",
        "height": undefined,
        "justifyContent": "flex-start",
        "marginRight": -24,
        "paddingBottom": 12,
        "paddingLeft": 0,
        "paddingTop": 16,
        "width": undefined,
      }
    }
  />,
]
`;
